### PR TITLE
New completion: apt-mark

### PR DESCRIPTION
**Pull request by JCGoran**
Mostly borrowed code from `apt-get` completion. I've also added some basic tests for edge case options.
Note that I've implemented the completion to display short options as well, which I could of course remove if they are deemed unnecessary.

### CODE REVIEW

**Comment by scop** [comment]
I'd rather see all `apt-*` completions being shipped with apt itself, as the `apt` completion is already there and IMO in the big picture it's better to have completions shipped along with the commands. I'm not "ruling out" inclusion of this in bash-completion in the meantime, as there is no active effort in getting them moved, and adding bits like this don't make it much harder when it actually starts to take place.

Will have a look at the actual PR content later unless someone else beats me to it.

**Comment by JCGoran** [comment]
Understood, though I don't see the completions being shipped with apt anytime soon, see [this bug on Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=799545), which was reported 5 years ago, and last activity was almost 3 years ago. Additionally, from what I can tell, `apt` itself ships only `apt` completion, while `bash-completion` offers the rest (`apt-get`, `apt-cache`, `apt-build`).

**Comment by scop** [comment]
Yeah, I've been involved with that bug. But I think the lack of response there should not be counted as a rejection. Someone just needs to do some work and then we'll see for real.

Anyway, addition of apt-mark here isn't a biggie, will have a quick look now.

File: test/t/test_apt_mark.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import pytest
+
+
+@pytest.mark.bashcomp(cmd="apt-mark")
+class TestAptMark:
+    @pytest.mark.complete("apt-mark ")
+    def test_1(self, completion):
+        assert all(x in completion for x in "auto manual remove showinstall showremove \
+        hold minimize-manual showauto showmanual unhold install \
```
Comment (scop) [inline test/t/test_apt_mark.py]: Not sure if it's just the patch viewer, or if we have a bit of an indentation issue here.

Anyway no need to use backslashes for line continuation, just use multiple string constants, they'll auto concatenate like
```python
... for x in (
    "auto manual remove showinstall showremove "
    "hold minimize-manual showauto showmanual unhold install "
    # [...]
)

File: test/t/test_apt_mark.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        purge showhold showpurge".split())
+
+    @pytest.mark.complete("apt-mark minimize-manual ")
+    def test_2(self, completion):
+        assert completion == []
```
Comment (scop) [inline test/t/test_apt_mark.py]: `assert not completion`

File: test/t/test_apt_mark.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    def test_3(self, completion):
+        assert completion == "--file=bash-completion-test-subject.deb"
+
+    @pytest.mark.complete("apt-mark --config-file ", cwd="dpkg")
+    def test_4(self, completion):
+        assert completion == "bash-completion-test-subject.deb"
+
+    @pytest.mark.complete("apt-mark --option ")
+    def test_5(self, completion):
+        assert completion == []
```
Comment (scop) [inline test/t/test_apt_mark.py]: `assert not completion`

File: completions/apt-mark [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        esac
+        return
+    fi
+
+    case $prev in
+        --help|--version|--option|-!(-*)[hvo])
+            return
+            ;;
+        --config-file|--file|-!(-*)[cf])
+            _filedir
```
Comment (scop) [inline completions/apt-mark]: Do you thing we could add some file extensions to narrow down the completions here? `*.conf` would probably work well enough for configs (I see we don't do that for apt-get either for reasons unknown to me though). Do the stats files accepted by `--file` follow a naming pattern we could use?

File: completions/apt-mark [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            return
+            ;;
+    esac
+
+    $split && return
+
+    if [[ "$cur" == -* ]]; then
+        COMPREPLY=( $(compgen -W '-f --file= --help --version -c --config-file -o --option' -- "$cur") )
```
Comment (scop) [inline completions/apt-mark]: Include only the long options here, search for "available options" in CONTRIBUTING.md

File: completions/apt-mark [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# Debian apt-mark(8) completion                             -*- shell-script -*-
+
+_apt_mark()
+{
+    local cur prev words cword split
+    _init_completion -s || return
+
+    local special i
+    for (( i=0; i < ${#words[@]}-1; i++ )); do
+        if [[ ${words[i]} == @(hold|unhold|auto|manual|install|remove|purge|showhold|showauto|showmanual|showinstall|showremove|showpurge|minimize-manual) ]]; then
```
Comment (scop) [inline completions/apt-mark]: This list would be better sorted correspondingly with the source for easier mainenance/crosscheck. The only source for the full list I'm aware of is the manpage, and at least in the version I have (1.6.12 on Ubuntu 18.04) they're not in this order.

My version doesn't have `minimize-manual`, can you confirm that its inclusion is intentional?

File: completions/apt-mark [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            ;;
+    esac
+
+    $split && return
+
+    if [[ "$cur" == -* ]]; then
+        COMPREPLY=( $(compgen -W '-f --file= --help --version -c --config-file -o --option' -- "$cur") )
+        [[ $COMPREPLY == *= ]] && compopt -o nospace
+    else
+        COMPREPLY=( $(compgen -W 'hold unhold install remove purge
```
Comment (scop) [inline completions/apt-mark]: Options list order and minimize-manual note from above applies here too.

File: completions/apt-mark [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+_apt_mark()
+{
+    local cur prev words cword split
+    _init_completion -s || return
+
+    local special i
+    for (( i=0; i < ${#words[@]}-1; i++ )); do
+        if [[ ${words[i]} == @(hold|unhold|auto|manual|install|remove|purge|showhold|showauto|showmanual|showinstall|showremove|showpurge|minimize-manual) ]]; then
+            special=${words[i]}
```
Comment (scop) [inline completions/apt-mark]: Should we `break` after this?

File: completions/apt-mark [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    done
+
+    if [[ -n $special ]]; then
+        case $special in
+            minimize-manual)
+                return
+                ;;
+            *)
+                COMPREPLY=( $(apt-cache --no-generate pkgnames "$cur" \
+                    2>/dev/null) )
```
Comment (scop) [inline completions/apt-mark]: Should reuse apt-cache's `_apt_cache_packages`, see examples in d90fbe3bfd7e53c365897fa584edc9d2d0a2e9bf which I just pushed.

File: completions/apt-mark [hunk 10]
```diff
@@ -0,0 +1,52 @@
+# Debian apt-mark(8) completion                             -*- shell-script -*-
+
+_apt_mark()
+{
+    local cur prev words cword split
+    _init_completion -s || return
+
+    local special i
+    for (( i=0; i < ${#words[@]}-1; i++ )); do
```
Comment (scop) [inline completions/apt-mark]: Could start from `i=1` (I'll fix this in a bunch of other completions we have)

File: completions/apt-mark (hunk 4 above)
Comment (JCGoran) [reply completions/apt-mark]: I've added `*.conf` in the latest version for `--config-file`, but for `--file=` I'm not 100% sure if there are any extensions, since by default it seems to read the file `/var/lib/apt/extended_states` which doesn't have one.

File: completions/apt-mark (hunk 8 above)
Comment (JCGoran) [reply completions/apt-mark]: Actually, since using more than one special option results in an error, here's a potential solution using an array, effectively disabling the completion if the user inputs more than one option:
```diff
diff --git a/completions/apt-mark b/completions/apt-mark
index a88aeacd..213ab6a4 100644
--- a/completions/apt-mark
+++ b/completions/apt-mark
@@ -5,15 +5,19 @@ _apt_mark()
     local cur prev words cword split
     _init_completion -s || return
 
-    local special i
+    local i
+    local special=()
     for (( i=1; i < ${#words[@]}-1; i++ )); do
         if [[ ${words[i]} == @(auto|manual|minimize-manual|showauto|showmanual|hold|unhold|showhold|install|remove|purge|showinstall|showremove|showpurge) ]]; then
-            special=${words[i]}
+            special+=(${words[i]})
         fi
     done
+    if [[ ${#special[@]} -gt 1 ]]; then
+        return
+    fi
 
-    if [[ -n $special ]]; then
-        case $special in
+    if [[ -n ${special[0]} ]]; then
+        case ${special[0]} in
             minimize-manual)
                 return
                 ;;
```

Comment (scop) [reply completions/apt-mark:11]: > Actually, since using more than one special option results in an error,

True, but:
```
# apt-mark unhold hold
E: Unable to locate package hold
E: No packages found
```
So the two specials are not actually treated as specials, the latter is treated as a package name. It just happens I don't have a package named "hold" installed, that's why the error. I don't think we should prevent cases like that from happening, so I opted for the simple `break`.

**Comment by scop** [comment]
Thanks, and apologies for lettings this stall for so long.
